Firestore/Source/Core/FSTView.h

```diff
@@ -108,7 +106,14 @@ typedef NS_ENUM(NSInteger, FSTLimboDocumentChangeType) {
  * @param docChanges The doc changes to apply to this view.
  * @return a new set of docs, changes, and refill flag.
  */
-- (FSTViewDocumentChanges *)computeChangesWithDocuments:(FSTMaybeDocumentDictionary *)docChanges;
+- (FSTViewDocumentChanges *)computeChangesWithDocuments:
+    (const firebase::firestore::model::DocumentMap &)docChanges;
+
+/** Like `computeChangesWithDocuments:`, but accepts a map of `MaybeDocument`s. */
+// PORTING NOTE: in C++, a container of `Derived` does not convert to a container of `Base`,
```

Comment:
@wilhuff Most of this PR is straightforward. The only thing I'd like to know your stance on is the issue that in C++, `SortedMap<Key, Derived*>` does not convert to `SortedMap<Key, Base*>` -- specifically, a `DocumentMap` cannot be passed in place of `MaybeDocumentMap`. I'd prefer to avoid creating a new map on the fly, due to performance concerns. The approach here is to have one method taking a `DocumentMap` and another taking a `MaybeDocumentMap`, and in the implementation file they both delegate to a template function. What do you think?

Reply:
I agree with the source of the problem--that in C++ different template arguments create unrelated types--but I wish we could find a way to avoid this altogether rather than accommodate the two through localized template magic.

Two thoughts:
  * Can we avoid needing both `DocumentMap` and `MaybeDocumentMap`?
  * Can we make these types non-aliases that do have an inheritance relationship so that these types are substitutable?

As I remember things, `DocumentMap` exists to make some cases more type-safe so that callers don't have to handle the non-Document results. However, that choice was made when all the languages we used allowed for covariant template arguments. I think we could reasonably revisit this rather than doubling the API surface area of anything that currently takes a `DocumentMap`.

Reply:
As I think about this more I'm convinced that we don't need `DocumentMap` to be a substitutable type for `MaybeDocumentMap`.

While it's true that `ExecuteQuery` should not produce anything other than Documents, we only use the result of `ExecuteQuery` to compute changes to views and that accepts MaybeDocuments. In a world where a `DocumentMap` can be cheaply substituted for `MaybeDocumentMap` the type safety this affords is nice, but isn't worth much to preserve.

On the flip side, having to return `DocumentMap` forces the implementation of `ExecuteQuery` to explicitly handle other types to prevent adding them to the query results. This is a very useful property that I don't want to give away.

I think what this means is that DocumentMap should have a has-a relationship with an underlying MaybeDocumentMap. You should not be able to construct a `DocumentMap` from a `MaybeDocumentMap` but you can build up a `DocumentMap` using APIs that only accept Documents and then ask for a cheap reference to the underlying MaybeDocumentMap. SortedMap is immutable so this should be cheap.

I think it should be sufficient for DocumentMap to be a very limited subset of SortedMap (omitting iteration, etc). WDYT?

Reply:
Let's proceed without the iterator. The "No iterator?" commit shows that this isn't really worth it.